charset/encoding.go

```diff
@@ -58,80 +59,94 @@ func (e *Encoding) Name() string {
+	return e.transform(e.enc.NewEncoder(), dest, src, false)
 }
 
 // Decode decodes the bytes to a string.
-func (e *Encoding) Decode(src []byte) (string, bool) {
-	return e.transform(e.enc.NewDecoder(), src)
+func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
+	return e.transform(e.enc.NewDecoder(), dest, src, true)
 }
 
-func (e *Encoding) transform(transformer transform.Transformer, src []byte) (string, bool) {
-	if len(e.buffer) < len(src) {
-		e.buffer = make([]byte, len(src)*2)
+func (e *Encoding) transform(transformer transform.Transformer, dest, src []byte, isDecoding bool) ([]byte, error) {
+	if len(dest) < len(src) {
+		dest = make([]byte, len(src)*2)
 	}
 	var destOffset, srcOffset int
-	ok := true
+	var encodingErr error
 	for {
-		nextLen := 4
-		if e.charLength != nil {
-			nextLen = e.charLength(src[srcOffset:])
-		}
-		srcEnd := srcOffset + nextLen
-		if srcEnd > len(src) {
-			srcEnd = len(src)
+		srcNextLen := e.nextCharLenInSrc(src[srcOffset:], isDecoding)
+		srcEnd := mathutil.Min(srcOffset+srcNextLen, len(src))
+		nDest, nSrc, err := transformer.Transform(dest[destOffset:], src[srcOffset:srcEnd], false)
+		if err == transform.ErrShortDst {
+			dest = enlargeCapacity(dest)
+		} else if err != nil || isDecoding && beginWithReplacementChar(dest[destOffset:destOffset+nDest]) {
+			if encodingErr == nil {
+				encodingErr = e.generateErr(src[srcOffset:], srcNextLen)
+			}
+			dest[destOffset] = byte('?')
+			nDest, nSrc = 1, srcNextLen // skip the source bytes that cannot be decoded normally.
 		}
-		nDest, nSrc, err := transformer.Transform(e.buffer[destOffset:], src[srcOffset:srcEnd], false)
 		destOffset += nDest
 		srcOffset += nSrc
-		if err == nil {
-			if srcOffset >= len(src) {
-				result := string(e.buffer[:destOffset])
-				if len(e.buffer) > encodingBufferSizeRecycleThreshold {
-					// This prevents Encoding from holding too much memory.
-					e.buffer = make([]byte, encodingBufferSizeDefault)
-				}
-				return result, ok
-			}
-		} else if err == transform.ErrShortDst {
-			newDest := make([]byte, len(e.buffer)*2)
-			copy(newDest, e.buffer)
-			e.buffer = newDest
-		} else {
-			e.buffer[destOffset] = byte('?')
-			destOffset += 1
-			srcOffset += 1
-			ok = false
+		// The source bytes are exhausted.
+		if srcOffset >= len(src) {
+			return dest[:destOffset], encodingErr
 		}
 	}
 }
+
+func (e *Encoding) nextCharLenInSrc(srcRest []byte, isDecoding bool) int {
+	if isDecoding && e.charLength != nil {
+		return e.charLength(srcRest)
+	}
+	return len(srcRest)
+}
+
+func enlargeCapacity(dest []byte) []byte {
+	newDest := make([]byte, len(dest)*2)
+	copy(newDest, dest)
+	return newDest
+}
+
+func (e *Encoding) generateErr(srcRest []byte, srcNextLen int) error {
+	cutEnd := mathutil.Min(srcNextLen, len(srcRest))
+	invalidBytes := fmt.Sprintf("%X", string(srcRest[:cutEnd]))
+	return errInvalidCharacterString.GenWithStackByArgs(e.name, invalidBytes)
+}
+
+var replacementBytes = []byte{0xEF, 0xBF, 0xBD}
```

Comment:
does this represent rune `0xfffd` in utf8? add some comments is good to understand

Reply:
+1 hard to understand

Reply:
Updated.